programs/marinade-finance/src/instructions/user/withdraw_stake_account.rs

```diff
@@ -0,0 +1,303 @@
+            MarinadeError::RequiredActiveStake
+        );
+
+        let mut validator = self
+            .state
+            .validator_system
+            .get(&self.validator_list.data.as_ref().borrow(), validator_index)?;
+
+        // check currently_staked in this account & validator vote-key
+        check_stake_amount_and_validator(
+            &self.stake_account,
+            stake.last_update_delegated_lamports,
+            &validator.validator_account,
+        )?;
+
+        // compute how many lamport the burned mSOL represents
+        let sol_value = self.state.msol_to_sol(msol_amount)?;
+        require_gt!(sol_value, self.state.min_withdraw);
+
+        // apply withdraw_stake_account_fee to avoid economical attacks
+        // withdraw_stake_account_fee must be >= one epoch staking rewards
+        let withdraw_stake_account_fee_lamports =
+            self.state.withdraw_stake_account_fee.apply(sol_value);
+        // the fee value will be burned but not delivered, thus increasing mSOL value slightly for all mSOL holders
+        let split_lamports = sol_value - withdraw_stake_account_fee_lamports;
+        // check withdraw amount (stake) >= self.state.stake_system.min_stake
+        // and also >= self.state.min_withdraw
+        require_gte!(
+            split_lamports,
+            std::cmp::max(self.state.min_withdraw, self.state.stake_system.min_stake),
+            MarinadeError::WithdrawAmountIsTooLow
+        );
+        // require remainder stake also >= self.state.stake_system.min_stake
+        // To simplify the flow, we always deliver the lamports in the splitted account,
+        // so some lamports must remain in the original account. Check that
+        // after split, the amount remaining in the stake account is >= state.stake_system.min_stake
+        require_gte!(
+            stake
+                .last_update_delegated_lamports
+                .saturating_sub(split_lamports),
+            self.state.stake_system.min_stake,
+            MarinadeError::StakeAccountRemainderTooLow
+        );
+
+        // burn mSOL
+        burn(
+            CpiContext::new(
+                self.token_program.to_account_info(),
+                Burn {
+                    mint: self.msol_mint.to_account_info(),
+                    from: self.burn_msol_from.to_account_info(),
+                    authority: self.burn_msol_authority.to_account_info(),
+                },
+            ),
+            msol_amount,
+        )?;
+        self.state.on_msol_burn(msol_amount)?;
+
+        // split split_lamports from stake account into out split_stake_account
+        msg!(
+            "Split {} lamports from stake {} into {}",
+            split_lamports,
+            stake.stake_account,
+            self.split_stake_account.key(),
+        );
+
+        let split_instruction = stake::instruction::split(
+            self.stake_account.to_account_info().key,
+            self.stake_withdraw_authority.key,
```

Comment:
Don't know if is it working with stake_withdraw_authority (usually it is DEPOSIT auth) but even if so, invoke signed is using STAKE_DEPOSIT_SEED not withdraw one so the signature will not match the key you are requesting to be used as authority